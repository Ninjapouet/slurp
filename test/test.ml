
let shell fmt =
  let k s = fun k ->
    (* Fmt.pr "shell: %s@." s; *)
    Lwt_process.(with_process_in (shell s) k) in
  Fmt.kstr k fmt

let test path res =
  shell "curl -s http://localhost:8080/%s" path
    (fun pin ->
       match%lwt Lwt_io.read_line pin#stdout with
       | s when s = res ->
         Fmt.pr "%s OK@." path;
         Lwt.return_unit
       | _ ->
         Fmt.pr "%s KO@." path;
         Lwt.return_unit
       | exception e ->
         Fmt.pr "%s %a@." path Fmt.exn e;
         Lwt.return_unit)

let _ =
  Lwt_main.run begin
    let cohttp = "../cohttp/server.exe" in
    let service = "services/sum.cmxs" in
    let ic = Unix.open_process_args_in cohttp
        [|cohttp; "-s"; service; "-p"; "8080"|]in
    Lwt_unix.sleep 1.;%lwt

    test "sum/2/3" "5";%lwt

    let pid = Unix.process_in_pid ic in
    Unix.kill pid (-9);
    Lwt.return_unit
  end
