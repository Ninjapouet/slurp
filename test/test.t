Launch the server.
  $ ./../bin/slurpd_cohttp.exe -s services/sum.cmxs,services/sum_lwt.cmxs,services/json.cmxs -p 8080 --static data > _log 2>&1 &
  $ echo $! > _pid
Wait a bit to let the server start...
  $ sleep 1
  $ cat _log

Test sum:
  $ curl -s --show-error http://localhost:8080/sum/2/3
  5

Test sum_lwt:
  $ curl -s --show-error http://localhost:8080/sum_lwt/2/3
  5

Test static file:
  $ curl -s --show-error http://localhost:8080/static/hello.txt
  Hello

Test static file with directory:
  $ curl -s --show-error http://localhost:8080/static/other/world.txt
  World

Test json data:
  $ curl -s --show-error http://localhost:8080/env/default
  {"foo":42,"bar":"pouet"}

Kill the server
  $ cat _pid | xargs kill -9
