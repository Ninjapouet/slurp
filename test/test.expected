sum/2/3 = 5 OK
sum_lwt/2/3 = 5 OK
static/hello.txt = Hello OK
static/other/world.txt = World OK
env/default = {"foo":42,"bar":"pouet"} OK
